channel/test/subscription.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"perun.network/go-perun/channel"
+)
+
+// SubscriptionTestSetup represents a setup for testing a subscription implementation.
+//
+// Subscription should return an instance of the subscription implementation.
+// EmitRegistered should make the subscription emit a registered event.
+// EmitProgressed should make the subscription emit a progressed event.
+// EmitConcluded should make the subscription emit a concluded event.
+type SubscriptionTestSetup interface {
+	Subscription() channel.AdjudicatorSubscription
+	EmitRegistered(context.Context) (channel.Params, channel.State)
+	EmitProgressed(context.Context) (channel.Params, channel.State)
+	EmitConcluded(context.Context) channel.Params
+}
+
+func TestSubscription(ctx context.Context, t *testing.T, s SubscriptionTestSetup) {
+	sub := s.Subscription()
+
+	{
```

Comment:
Will you put `t.Run` calls here or a comment?

Reply:
I didn't plan to do so, because the code is dependent on each other. You can't run the later parts without the earlier parts.